crates/compiler/src/obj_gen/util.rs

```diff
@@ -0,0 +1,265 @@
+            | InstructionOpcode::CatchRet
+            | InstructionOpcode::CleanupRet
+            | InstructionOpcode::Unreachable
+    )
+}
+
+/// Extracts the value of the provided `operand` as a basic value where
+/// possible.
+///
+/// # Errors
+///
+/// - [`Error::InvalidOpcodeOperand`] when the provided operand is a basic
+///   block.
+pub fn extract_value_operand(
+    operand: OptionalInkwellOperand,
+    opcode: InstructionOpcode,
+) -> Result<BasicValueEnum> {
+    match operand {
+        Some(Either::Left(basic_value)) => Ok(basic_value),
+        Some(Either::Right(basic_block)) => {
+            let block_name = basic_block.get_name().to_str()?;
+            Err(Error::invalid_opcode_operand(
+                opcode,
+                &format!("A value was expected where basic block {block_name} was encountered"),
+            ))
+        }
+        None => Err(Error::invalid_opcode_operand(
+            opcode,
+            "No operand was found where one was required",
+        )),
+    }
+}
+
+/// Extracts the value of the provided `operand` as a basic block where
+/// possible.
+///
+/// # Errors
+///
+/// - [`Error::InvalidOpcodeOperand`] when the provided operand is a basic
+///   value.
+pub fn extract_block_operand(
+    operand: OptionalInkwellOperand,
+    opcode: InstructionOpcode,
+) -> Result<BasicBlock> {
+    match operand {
+        Some(Either::Right(basic_block)) => Ok(basic_block),
+        Some(Either::Left(basic_value)) => {
+            let value_name = basic_value.get_name().to_str()?;
+            Err(Error::invalid_opcode_operand(
+                opcode,
+                &format!("A value was expected where basic value {value_name} was encountered"),
+            ))
+        }
+        None => Err(Error::invalid_opcode_operand(
+            opcode,
+            "No operand was found where one was required",
+        )),
+    }
+}
+
+/// Extracts the operands to the provided `instruction` while asserting that
+/// they are all basic values (variants of [`BasicValueEnum`]).
+///
+/// # Errors
+///
+/// - [`Error::InvalidOpcodeOperand`] when one of the operands to `instruction`
+///   is a basic block.
+pub fn extract_value_operands(instruction: InstructionValue) -> Result<Vec<BasicValueEnum>> {
+    let opcode = instruction.get_opcode();
+    instruction
+        .get_operands()
+        .map(|operand| extract_value_operand(operand, opcode))
+        .collect::<Result<Vec<BasicValueEnum>>>()
+}
+
+/// Extracts pairs of `(name, type)` from the provided value operands.
+///
+/// # Errors
+///
+/// - [`Error::UnsupportedType`] if a scalable vector value is encountered in
+///   the provided `operands`.
+pub fn name_type_pairs_from_value_operands(
+    operands: &[BasicValueEnum],
+) -> Result<Vec<(String, LLVMType)>> {
+    operands
+        .iter()
+        .map(|bv| match bv {
+            BasicValueEnum::ArrayValue(array_val) => {
+                let ty = LLVMType::try_from(array_val.get_type())?;
```

Comment:
nit: lots of duplication -- can this repeated segment be an inner function instead?

Reply:
No it can't. The types differ.